sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -1,20 +1,56 @@
-# Author: Andrew nystrom <awnystrom@gmail.com>
+# Authors: Andrew nystrom <awnystrom@gmail.com>
+#          Meekail Zain <zainmeekail@gmail.com>
+from libc.math cimport sqrt, pow
+cimport numpy as cnp
+cnp.import_array()
 
 from scipy.sparse import csr_matrix
-cimport numpy as cnp
 
-cnp.import_array()
-ctypedef cnp.int32_t INDEX_T
+from libc.limits cimport INT_MAX, LONG_MAX
+
+ctypedef cnp.int8_t FLAG_t
+
+# This is (approximately) the maximum value before the intermediate computation
+# i**2 + 3*i blows out.
+# (Approximate) solution to i**2 + 3*i = maxint32
+cdef cnp.int64_t MAX_SAFE_INDEX_DEG2 = <cnp.int64_t> sqrt(LONG_MAX) - 4
 
-ctypedef fused DATA_T:
+# This is (approximately) the maximum value before the intermediate computation
+# 3 * d**2 * d + d**3 blows out, since d is the upper bound of i.
+# (Approximate) solution to 3 * d**2 * d + d**3 = maxint32
+cdef cnp.int64_t MAX_SAFE_INDEX_DEG3 = <cnp.int64_t> pow(LONG_MAX, 1/3)/4
+
+# This is the maximum value before the intermediate computation
+# d**2 + d blows out
```

Comment:
I think "overflow" is a better phrase here.

```suggestion
# d**2 + d overflows
```

This comment applies in other places.